### PR TITLE
StateDB based Health

### PR DESCRIPTION
**Pull request by tommyp1ckles**
# healthv2: add new module health implementation based on statedb.

Similar to existing implementation in hive/cell/<health/structured>.go, this provides system health data in a tree structure.
However, this seeks to reimplement the health provider such that it is no longer coupled to pkg/hive/cell.
Because this will eventually be run just as any other Agent module, we can use StateDB for the implementation.

For more information about the original tree structured health reporter take a look at code documentation [here](https://github.com/cilium/cilium/blob/main/pkg/hive/cell/structured.go#L59-L94). This seeks to provide a similar underlying structure of data while using a much simpler statedb schema.

### Note: The only intended user impact of these changes is that health data will now be available with `cilium-dbg statedb health` and `cilium-dbg statedb dump`. 

This massively reduces the complexity of the health implementation, as well as provides a more convenient data model for storing health update data.

Furthermore, this fixes the awkward distinction between "module" and "subcomponent" reporting by unifying how these health reports are stored in one place.

pkg/healthv2/provider.go implements a new health provider that uses statedb to store a table of health updates by their fully qualified identifier. This is composed of:

[module-id].[componenet-id]

Where the module ID is a fully qualified ID of submodules, ex:

i.  agent.controlplane.bgpv1

Example `cilium statedb health`output:

```bash
Module                                           Component                                                                                              Level   Message                                                        LastOK                 UpdatedAt              Count
agent.controlplane.endpoint-manager              endpoint-gc                                                                                            OK      endpoint-gc                                                    2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.l2-announcer                  job-l2-announcer lease-gc                                                                              OK      Running                                                        2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.auth                          observer-job-auth gc-identity-events                                                                   OK      Primed                                                         2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.auth                          timer-job-auth gc-cleanup                                                                              OK      Primed                                                         2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.auth                          observer-job-auth request-authentication                                                               OK      Primed                                                         2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.node-manager                  nodes-add                                                                                              OK      Node adds successful                                           2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   3
agent.controlplane.node-manager                  nodes-update                                                                                           OK      Node updates successful                                        2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   1
agent.datapath.node-address                      job-node-address-update                                                                                OK      10.244.1.33 (cilium_host), fd00:10:244:1::3c26 (cilium_host)   2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.bgp-cp                        job-diffstore-events                                                                                   OK      Running                                                        2024-03-06T23:27:22Z   2024-03-06T23:27:22Z   1
agent.controlplane.stale-endpoint-cleanup        job-endpoint-cleanup                                                                                   OK      Running                                                        2024-03-06T23:27:22Z   2024-03-06T23:27:22Z   0
agent.controlplane.envoy-proxy                   timer-job-version-check                                                                                OK      Primed                                                         2024-03-06T23:27:22Z   2024-03-06T23:27:22Z   0
agent.controlplane.service-manager               job-ServiceReconciler                                                                                  OK      2 NodePort frontend addresses                                  2024-03-06T23:27:22Z   2024-03-06T23:27:22Z   0
agent.datapath.l2-responder                      job-l2-responder-reconciler                                                                            OK      Running                                                        2024-03-06T23:27:22Z   2024-03-06T23:27:22Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-1865 (/).policymap-sync                                                                OK      sync-policymap-1865                                            2024-03-06T23:27:23Z   2024-03-06T23:27:23Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-1865 (/).datapath-regenerate                                                           OK      Endpoint regeneration successful                               2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   2
agent.controlplane.endpoint-manager              cilium-endpoint-162 (local-path-storage/local-path-provisioner-7577fdbbfb-pcffg).policymap-sync        OK      sync-policymap-162                                             2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-162 (local-path-storage/local-path-provisioner-7577fdbbfb-pcffg).datapath-regenerate   OK      Endpoint regeneration successful                               2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   1
agent.controlplane.endpoint-manager              cilium-endpoint-2319 (kube-system/coredns-76f75df574-8m4d8).policymap-sync                             OK      sync-policymap-2319                                            2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-2319 (kube-system/coredns-76f75df574-8m4d8).datapath-regenerate                        OK      Endpoint regeneration successful                               2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   1
agent.controlplane.endpoint-manager              cilium-endpoint-605 (kube-system/coredns-76f75df574-ql8kj).policymap-sync                              OK      sync-policymap-605                                             2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-605 (kube-system/coredns-76f75df574-ql8kj).datapath-regenerate                         OK      Endpoint regeneration successful                               2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   1
agent.controlplane.endpoint-manager              cilium-endpoint-3399 (/).policymap-sync                                                                OK      sync-policymap-3399                                            2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-3399 (/).datapath-regenerate                                                           OK      Endpoint regeneration successful                               2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   1
agent.controlplane.node-manager                  background-sync                                                                                        OK      Node validation successful                                     2024-03-06T23:29:57Z   2024-03-06T23:29:57Z   2
agent.controlplane.daemon                        job-sync-hostips                                                                                       OK      Synchronized                                                   2024-03-06T23:30:22Z   2024-03-06T23:30:22Z   3
agent.controlplane.daemon.ep-bpf-prog-watchdog   ep-bpf-prog-watchdog                                                                                   OK      ep-bpf-prog-watchdog                                           2024-03-06T23:30:22Z   2024-03-06T23:30:22Z   6
agent.datapath.sysctl                            job-reconciler-loop                                                                                    OK      OK, 25 objects                                                 2024-03-06T23:30:41Z   2024-03-06T23:30:41Z   62
agent.controlplane.endpoint-manager              cilium-endpoint-2319 (kube-system/coredns-76f75df574-8m4d8).cep-k8s-sync                               OK      sync-to-k8s-ciliumendpoint (2319)                              2024-03-06T23:30:42Z   2024-03-06T23:30:42Z   20
agent.controlplane.endpoint-manager              cilium-endpoint-162 (local-path-storage/local-path-provisioner-7577fdbbfb-pcffg).cep-k8s-sync          OK      sync-to-k8s-ciliumendpoint (162)                               2024-03-06T23:30:42Z   2024-03-06T23:30:42Z   20
agent.controlplane.endpoint-manager              cilium-endpoint-605 (kube-system/coredns-76f75df574-ql8kj).cep-k8s-sync                                OK      sync-to-k8s-ciliumendpoint (605)                               2024-03-06T23:30:42Z   2024-03-06T23:30:42Z   20
agent.datapath.agent-liveness-updater            timer-job-agent-liveness-updater                                                                       OK      OK (107.75µs)                                                  2024-03-06T23:30:43Z   2024-03-06T23:30:43Z   0

```

Similarily the component id is a ID representing a tree of subcomponents of a module.  Together, these form a tree where each path stores information about the module and component being reported on.  As well, Status updates store the original "two-component" identifier.

This schema is meant to be less opinionated about how to view health data, it is simply a set of health report rows indexed by a identifier path.
Because of this, there is no longer any distinction between a "reporter" (i.e. leaf) and "scope" (i.e. parent node). This means that a reporter can have a status and have child reports.

Initially this will be shimmed into the existing health infrastructure in hive/cell. Ultimately we will remove all that code and refactor health reporters using the external github.com/cilium/hive library.

### CODE REVIEW

File: api/v1/openapi.yaml
Comment (learnitall) [inline api/v1/openapi.yaml]: Nit: the description here is formatted as a question:

> How many times in a row has this status been reported?

Can we reword this to be a statement?

```suggestion
        description: The number of consecutive status updates.
```

Another option:

```The number of times in a row that the status has been reported.```

**Review by learnitall — CHANGES_REQUESTED** [review]
For API I just have a small nit, otherwise LGTM.

File: pkg/healthv2/provider.go
Comment (joamaki) [inline pkg/healthv2/provider.go]: Let's make `statedb.Table[types.Status]` available to the whole application (also the indexes). This allows tests to check that the status for example.

File: pkg/healthv2/provider.go
Comment (joamaki) [inline pkg/healthv2/provider.go]: Prefer doing defer of the abort instead of commit. The log message that we upserted is also printed now in case of failed insert.

File: pkg/healthv2/cell.go
Comment (joestringer) [inline pkg/healthv2/cell.go:1]: Looks like these files need a CODEOWNERS update for ownership, presumably to sig-agent?

**Comment by joestringer** [comment]
For this to be ready for review, it would also help reviewers if the PR description provided an overview of the changes, including what (if any) user-facing changes that this will introduce.

File: pkg/healthv2/provider.go
Comment (dylandreimerink) [inline pkg/healthv2/provider.go]: Perhaps nicer/better to use the existing Strigger index helpers https://github.com/cilium/cilium/blob/main/pkg/statedb/index/string.go#L12

File: pkg/healthv2/provider.go
Comment (dylandreimerink) [inline pkg/healthv2/provider.go]: Shouldn't we want to use a FieldLogger provided by hive? Or pkg/logging ?

File: pkg/healthv2/provider.go
Comment (dylandreimerink) [inline pkg/healthv2/provider.go]: Isn't this easier?
```suggestion
			old, _, found := p.statusTable.First(tx, p.PrimaryIndex.QueryFromObject(s))
			// If a similar status already exists, increment count, otherwise start back
			// at zero.
			if found && old.Level == s.Level && old.Message == s.Message {
				s.Count = old.Count + 1
			}
```

**Review by dylandreimerink — CHANGES_REQUESTED** [review]

**Comment by sayboras** [comment]
remove myself as reviewer because vendor team is no longer required

File: pkg/healthv2/provider.go
Comment (joamaki) [inline pkg/healthv2/provider.go]: Here as well we can do `index.Stringer[types.HealthID]` (and perhaps Go will be able to infer the type in the future...)

**Review by joamaki — APPROVED** [review]

File: cilium-dbg/cmd/status.go
Comment (derailed) [inline cilium-dbg/cmd/status.go]: nit: we should probably beef this up to report additional context vs lowerbound failed

File: pkg/health/client/modules.go
Comment (derailed) [inline pkg/health/client/modules.go]: Delete?

**Review by derailed — COMMENTED** [review]
@tommyp1ckles Nice work!

File: pkg/healthv2/provider.go
Comment (joamaki) [reply pkg/healthv2/provider.go]: I think the idea was to have non-terminated key so prefix searches would work. We discussed with @tommyp1ckles about also changing index.String to not terminate and just rely on the iterators to cut off after length mismatches. Though we can also add e.g. index.StringPrefix which is like index.String but no termination and then use that for FromKey, but perhaps less confusing overall if we just drop the key termination.

File: cilium-dbg/cmd/status.go
Comment (tommyp1ckles) [reply cilium-dbg/cmd/status.go]: This was based on the cilium-dbg/cmd/statedb.go "statedb" dump code.  But I'll add a better message in case this ever fails.

File: pkg/healthv2/provider.go
Comment (tommyp1ckles) [reply pkg/healthv2/provider.go]: That makes sense 👍

File: pkg/healthv2/provider.go
Comment (tommyp1ckles) [reply pkg/healthv2/provider.go]: @dylandreimerink @joamaki yeah, I'd like to play around adding support for prefix querys, including partial key match. For example something like:

agent.control-plane.endpoint-manager.endpoint-* that might match a set of endpoint specific status such as:

 ```
agent.control-plane.endpoint-manager.endpoint-1234.k8s-sync
agent.control-plane.endpoint-manager.endpoint-1234.datapath-reconcile
agent.control-plane.endpoint-manager.endpoint-111.k8s-sync
agent.control-plane.endpoint-manager.endpoint-111.datapath-reconcile
...
```

File: pkg/healthv2/provider.go
Comment (tommyp1ckles) [reply pkg/healthv2/provider.go]: Yup, good catch, this was supposed to be a temporary measure while developing the code, ill switch that over.

File: CODEOWNERS
Comment (bimmlerd) [inline CODEOWNERS:461]: `healthv2` reads weird to me. For modules, it'd be `health/v2` (see [here](https://go.dev/blog/v2-go-modules)) but I'm not sure what the way to go is for packages in tree. Is the goal to eventually rename this `health` again?

File: cilium-dbg/cmd/status.go
Comment (bimmlerd) [inline cilium-dbg/cmd/status.go]: duplicate import? (`healthTypes` and `types`)

File: pkg/hive/cell/module.go
Comment (bimmlerd) [inline pkg/hive/cell/module.go]: duplicate import?

**Review by bimmlerd — APPROVED** [review]
CODEOWNERS is fine by me, couple of nits

File: CODEOWNERS
Comment (tommyp1ckles) [reply CODEOWNERS:461]: health/v2 is definitely cleaner, however there is already a `pkg/health` that has nothing to do with this code so that may be confusing to have that nested inside there.

Perhaps something like healthprovider may be better. pkg/healthv2 may also mislead someone into thinking this is related to pkg/health.

Comment (bimmlerd) [reply CODEOWNERS:461]: Oh! Yeah for some reason I definitely assumed this is version2 of the same pkg >.<

**Comment by tommyp1ckles** [comment]
@derailed @pippolo84 @learnitall @dylandreimerink Gentle ping for review 😄

File: cilium-dbg/cmd/status.go
Comment (derailed) [inline cilium-dbg/cmd/status.go]: Where is this used? Since we've already queried the db.

File: pkg/health/client/modules_test.go
Comment (derailed) [inline pkg/health/client/modules_test.go]: did we loose the sort order?

**Review by joamaki — CHANGES_REQUESTED** [review]

File: pkg/healthv2/provider.go
Comment (joamaki) [inline pkg/healthv2/provider.go]: I would not export `RWTable` but keep it private to the provider. 

I would also make the `Table[types.Status]` available separately from `*HealthProvider`, and also have the indexes as top-level variables instead embedded into the health provider. The idea being that we would try to not "hide" the read-only access to statedb tables as they're central to how components using them work. E.g. it's easier to understand the health stuff when you see the table and `types.Status`.

File: pkg/healthv2/provider.go
Comment (joamaki) [inline pkg/healthv2/provider.go]: For consistency with definitions elsewhere we should call this `HealthTableName`.

File: pkg/healthv2/cell.go
Comment (joamaki) [inline pkg/healthv2/cell.go:4]: I think we could move this into `pkg/hive/health` once the relocation of `pkg/hive` to `cilium/hive` is done. We'll anyway have `pkg/hive` that has the `cilium/cilium` specific wrapper for `hive.New` etc. That wrapper would then by default include stuff like health, statedb and so on.

File: pkg/healthv2/types/types.go
Comment (derailed) [inline pkg/healthv2/types/types.go:26]: nit missing docs on most exports in this file

File: pkg/statedb/reconciler/reconciler_test.go
Comment (derailed) [inline pkg/statedb/reconciler/reconciler_test.go]: nit: axe?

File: pkg/statedb/reconciler/reconciler_test.go
Comment (derailed) [inline pkg/statedb/reconciler/reconciler_test.go]: nit: axe?

**Review by derailed — COMMENTED** [review]
@tommyp1ckles Thanks for the updates!

**Review by learnitall — APPROVED** [review]
Looks good on docs, thanks! Up to you, but I think it would be a good idea to add a quick note in the upgrade notes about the breaking changes in `cilium-dbg` usage.

**Review by pippolo84 — APPROVED** [review]
Operator changes LGTM, thanks :100:

File: pkg/health/client/modules_test.go
Comment (tommyp1ckles) [reply pkg/health/client/modules_test.go]: This might be related to how we query the db, let me take another look at this.

Comment (tommyp1ckles) [reply pkg/health/client/modules_test.go]: Changed up how we do the remote statedb query, which will result in the status' to be sent in alphabetical order (sorted by the identifier key).  

~Do you think the formatting code also ensure that the output is sorted?  Currently the tree will be built relative to the order in which the the Status slice is passed.~

@derailed Fixed this to sort by identifier prior to building the tree, this may be a bit redundant but it'll keep tests and output consistent.

**Review by dylandreimerink — APPROVED** [review]

File: pkg/healthv2/types/types.go
Comment (tommyp1ckles) [reply pkg/healthv2/types/types.go:26]: Add docs for relevant exported vars (skipping things like implementation of fmt.Stringer etc).

**Review by qmonnet — CHANGES_REQUESTED** [review]
> Looks good on docs, thanks! Up to you, but I think it would be a good idea to add a quick note in the upgrade notes about the breaking changes in `cilium-dbg` usage.

+1, please do add a note. Doc changes look good otherwise.

**Review by derailed — APPROVED** [review]
@tommyp1ckles Thanks for the updates!

**Review by qmonnet — APPROVED** [review]
Thanks!

**Review by lmb — APPROVED** [review]

**Pull request merged**
